plugins/sharding/logical_count.rb

```diff
@@ -57,79 +42,296 @@ def cache_key(input)
+          @result_sets = []
+          @temporary_tables_queue = []
+          @time_classify_types = detect_time_classify_types
+          @referred_objects_queue = []
+        end
+
+        def temporary_tables
+          @temporary_tables_queue.last
+        end
+
+        def referred_objects
+          @referred_objects_queue.last
+        end
+
+        def push
+          @temporary_tables_queue << []
+          @referred_objects_queue << []
+        end
+
+        def shift
+          temporary_tables = @temporary_tables_queue.shift
+          temporary_tables.each(&:close)
+          temporary_tables.clear
+          referred_objects = @referred_objects_queue.shift
+          referred_objects.each(&:unref)
+          referred_objects.clear
+        end
+
+        def close
+          until @temporary_tables_queue.empty?
+            shift
+          end
+          @result_sets.clear
+          @enumerator.unref
+        end
+
+        def consume_result_sets
+          while (result_set = @result_sets.shift)
+            yield(result_set)
+          end
+        end
+
+        def need_look_ahead?
+          return false unless @dynamic_columns.have_window_function?
+          return false unless @time_classify_types.empty?
+          true
+        end
+
+        def detect_time_classify_types
+          window_group_keys = []
+          @dynamic_columns.each_filtered do |dynamic_column|
+            window_group_keys.concat(dynamic_column.window_group_keys)
+          end
+          return [] if window_group_keys.empty?
+
+          types = []
+          @dynamic_columns.each do |dynamic_column|
+            next unless window_group_keys.include?(dynamic_column.label)
+            case dynamic_column.value.strip
+            when /\Atime_classify_(.+?)\s*\(\s*([a-zA-Z\d_]+)\s*[,)]/
+              type = $1
+              column = $2
+              next if column != @enumerator.shard_key_name
+
+              case type
+              when "minute", "second"
+                types << type
+              when "day", "hour"
+                types << type
+              end
+            end
+          end
+          types
         end
       end
 
       class Counter
+        def initialize(context)
+          @context = context
+        end
+
+        def count
+          have_shard = false
+          have_result_set = false
+          total = 0
+          each_shard_executor do |shard_executor|
+            have_shard = true
+            total += shard_executor.count
+          end
+          unless have_shard
+            enumerator = @context.enumerator
+            message =
+              "[logical_count] no shard exists: " +
+              "logical_table: <#{enumerator.logical_table}>: " +
+              "shard_key: <#{enumerator.shard_key_name}>"
+            raise InvalidArgument, message
+          end
+          total
+        end
+
+        private
+        def each_shard_executor(&block)
```

Comment:
This is similar to `Groonga::Sharding::LogicalRangeFilterCommand::Executer#each_shard_executor`.
Can we unify common implementation?

Reply:
I have unified common implementation with `logical_range_filter` in #1342, #1343, #1345.